apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/backend/ir/BackendIrBuilder.kt

```diff
@@ -1,205 +1,298 @@
+        typeDeclarations = input.typeDeclarations.map {
+          IntrospectionSchema.TypeRef(
+              kind = it.schemaKind(),
+              name = it.name
+          )
+        },
+        typesPackageName = input.typesPackageName,
+        fragmentsPackageName = input.fragmentsPackageName,
     )
   }
 
-  private fun Operation.buildBackendIrOperation(): BackendIr.Operation {
+  private fun GQLOperationDefinition.buildBackendIrOperation(): BackendIr.Operation {
     val normalizedName = this.normalizeOperationName()
-    val typeRef = this.operationTypeRef
+    val rootTypeDefinition = this.rootTypeDefinition(schema)!!
     val selectionKey = SelectionKey(
         root = normalizedName,
         keys = listOf(normalizedName, "data"),
         type = SelectionKey.Type.Query,
     )
-    val dataField = Field(
-        fieldName = "data",
-        responseName = "data",
-        type = typeRef.name!!,
-        fields = this.fields,
-        inlineFragments = this.inlineFragments,
-        fragmentRefs = this.fragments,
-        typeDescription = "",
-        sourceLocation = SourceLocation.UNKNOWN
+    val dataFieldDefinition = GQLFieldDefinition(
+        description = "",
+        name = "data",
+        arguments = emptyList(),
+        directives = emptyList(),
+        type = GQLNamedType(name = rootTypeDefinition.name)
+    )
+    val dataField = GQLField(
+        name = "data",
+        alias = null,
+        sourceLocation = com.apollographql.apollo.compiler.frontend.gql.SourceLocation.UNKNOWN,
+        arguments = null,
+        directives = emptyList(),
+        selectionSet = selectionSet
     ).buildBackendIrField(
         selectionKey = selectionKey,
         generateFragmentImplementations = true,
+        dataFieldDefinition
     )
-    val variables = this.variables.map { variable ->
+    val variables = this.variableDefinitions.map { variable ->
       BackendIr.Variable(
           name = variable.name,
-          type = schema.resolveType(variable.type)
+          type = variable.type.toSchemaType(schema)
       )
     }
+    val fragmentNames = usedFragmentNames(schema, fragmentDefinitions)
     return BackendIr.Operation(
         name = normalizedName,
-        operationName = this.operationName,
-        targetPackageName = this.packageName,
-        operationType = typeRef,
-        comment = this.description,
+        operationName = name!!,
+        targetPackageName = packageNameProvider?.operationPackageName(filePath = sourceLocation.filePath ?: "") ?: "",
+        operationType = IntrospectionSchema.TypeRef(
+            IntrospectionSchema.Kind.OBJECT,
+            rootTypeDefinition.name
+        ),
+        comment = this.description ?: "",
         variables = variables,
-        definition = this.sourceWithFragments,
+        definition = (toUtf8WithIndents() + "\n" + fragmentNames.map { fragmentDefinitions[it]!!.toUtf8WithIndents() }.joinToString("\n")).trimEnd('\n'),
```

Comment:
nit: I think it can be shortent  
```
toUtf8WithIndents() + fragmentNames.joinToString(
  separator = "\n"
) { fragmentName ->
  fragmentDefinitions[fragmentName]!!.toUtf8WithIndents()
}
```